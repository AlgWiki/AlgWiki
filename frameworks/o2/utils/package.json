{
  "name": "@oxy2/utils",
  "version": "0.0.1",
  "description": "Internal utilities for the O2 framework.",
  "keywords": [],
  "homepage": "https://github.com/AlgWiki/AlgWiki#readme",
  "repository": {
    "type": "git",
    "url": "https://github.com/AlgWiki/AlgWiki.git",
    "directory": "frameworks/o2/utils"
  },
  "bugs": {
    "url": "https://github.com/AlgWiki/AlgWiki/issues"
  },
  "author": "jakzo",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "engines": {
    "node": ">=14"
  },
  "devDependencies": {
    "@types/node": "14.14.37"
  },
  "dependencies": {
    "tslib": "^2.2.0"
  }
}
