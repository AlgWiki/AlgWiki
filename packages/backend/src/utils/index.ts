export * from './context';
export * from './logger';
