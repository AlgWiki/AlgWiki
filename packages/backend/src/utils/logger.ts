import pino from 'pino';

export const baseLogger = pino({ base: null });
