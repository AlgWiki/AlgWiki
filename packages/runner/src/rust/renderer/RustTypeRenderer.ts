import { Primitive, Renderer } from "../../Type";

export class RustTypeRenderer implements Renderer {
  single(t: Primitive): string {
    switch (t) {
      case Primitive.Integer:
        return "i64";
      case Primitive.Float:
        return "f64";
      case Primitive.Boolean:
        return "bool";
      case Primitive.String:
        return "&str";
    }
  }

  list(t: Primitive): string {
    return `Vec<${this.single(t)}>`;
  }

  linkedList(t: Primitive): string {
    return `LinkedList<${this.single(t)}>`;
  }

  dictionary(t: { key: Primitive; value: Primitive }): string {
    return `HashMap<${this.single(t.key)}, ${this.single(t.value)}>`;
  }
}
