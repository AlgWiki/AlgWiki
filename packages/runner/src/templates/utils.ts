import { Language } from "@alg-wiki/types";
import { snakeCase } from "change-case";

import { RustTypeRenderer } from "./rust";
import { TypeRenderer } from "./types";

/**
 * A map of `TypeRenderer`s for each language (if any)
 */
export const LanguageTypeRenderer: Record<Language, TypeRenderer | null> = {
  [Language.Python3]: null,
  [Language.Rust]: new RustTypeRenderer(),
};

/**
 * Functions to create a valid identifier for each language
 * TODO: need to ensure these functions exclude language keywords, etc
 */
export const LanguageIdentifier: Record<Language, (inp: string) => string> = {
  [Language.Rust]: (inp) =>
    `${/[^a-z]/.test(inp[0]) ? "_" : ""}${snakeCase(inp)}`,
  [Language.Python3]: (inp) =>
    `${/[^a-z]/.test(inp[0]) ? "_" : ""}${snakeCase(inp)}`,
};

/**
 * The filename for the file containing the runner, which will be created in the
 * bind mount directory
 */
export const LanguageFileName: Record<Language, string> = {
  [Language.Python3]: "index.py",
  [Language.Rust]: "main.rs",
};
