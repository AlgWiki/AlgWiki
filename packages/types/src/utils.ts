import { snakeCase } from "change-case";

import { EncodedResultError, Language } from "./types";

export function isObject(x: unknown): x is Record<string, unknown> {
  // If `x` is already an object, it's returned. Otherwise it's wrapped.
  // Therefore this will check if a value is an object, and also rule out null/undefined.
  return x === new Object(x);
}

// TODO: doesn't perform a deep check
export function isEncodedResultError(
  x: unknown,
  boundaryKey: string
): x is EncodedResultError {
  return isObject(x) && boundaryKey in x;
}

export const LanguageCasing: Record<Language, (name: string) => string> = {
  [Language.Python3]: snakeCase,
};
