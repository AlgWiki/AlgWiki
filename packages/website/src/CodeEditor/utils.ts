export const isCodeMirrorRequired = (language: string): boolean => {
  const isMobile = false;
  const isLanguageSupportedByMonaco = language === "javascript";
  return isMobile || !isLanguageSupportedByMonaco;
};
