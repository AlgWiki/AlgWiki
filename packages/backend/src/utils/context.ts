import { createNamespace } from 'cls-hooked';
import { Logger } from 'pino';
import { baseLogger } from './logger';

export interface RequestContext {
  log: Logger;
}

export const ctxNs = createNamespace('request');

export const runWithCtx = <T>(ctx: RequestContext, cb: () => T) =>
  ctxNs.runAndReturn(() => {
    ctxNs.active.ctx = ctx;
    return cb();
  });

export const getCtx = () => ctxNs.active.ctx as RequestContext | undefined;

export const defaultCtx: RequestContext = {
  log: baseLogger,
};

/** Request context object with defaults when context has not been initialized. */
export const ctx = new Proxy(defaultCtx, {
  get(target, key) {
    const reqCtx = getCtx() || target;
    return (reqCtx as any)[key];
  },
});
