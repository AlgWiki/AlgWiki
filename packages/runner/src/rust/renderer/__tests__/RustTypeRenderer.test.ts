import { RustTypeRenderer } from "../RustTypeRenderer";
import { typeMatrix } from "./utils";

describe("RustTypeRenderer", () => {
  it("can render all types", () => {
    const renderedTypes = typeMatrix
      .map((type) => type.render(new RustTypeRenderer()))
      .join("\n");
    expect(renderedTypes).toMatchSnapshot();
  });
});
