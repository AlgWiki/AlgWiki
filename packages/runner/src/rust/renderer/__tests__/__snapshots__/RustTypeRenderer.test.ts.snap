// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RustTypeRenderer can render all types 1`] = `
"i64
Vec<i64>
LinkedList<i64>
HashMap<i64, f64>
HashMap<i64, &str>
HashMap<i64, bool>
f64
Vec<f64>
LinkedList<f64>
HashMap<f64, i64>
HashMap<f64, &str>
HashMap<f64, bool>
&str
Vec<&str>
LinkedList<&str>
HashMap<&str, i64>
HashMap<&str, f64>
HashMap<&str, bool>
bool
Vec<bool>
LinkedList<bool>
HashMap<bool, i64>
HashMap<bool, f64>
HashMap<bool, &str>"
`;
